### DISCUSSION TITLE
Receive-waiting methods like spin_once and spin_some in ROS

### DISCUSSION THREAD

**Discussion by akihiro-komorilab** [eclipse-zenoh/roadmap#109] [Q&A]
Hi,

In a single-threaded program, I would to subscribe to a message and do some other processing, like spin_once() in rclpy or spin_some() in rclcpp. How should I implement this in zenoh-python?

Thanks,

**Comment by Mallets** [comment]
`zenoh-python` is based on the Rust implementation which is already multi-threaded and asynchronous. If you want to be single-threaded, e.g. on embedded targets, you could use [zenoh-pico](https://github.com/eclipse-zenoh/zenoh-pico) (but it's in C).

**Reply by akihiro-komorilab** [reply]
Thank you for your answer.
Is there any API to join() that thread and wait for it to exit?

**Reply by Mallets** [reply]
Those thread are handled by Rust and used internally. Therefore, they are not available to use by the application.
Would you mind explaining a bit more about your use case and what are you trying to achieve? Maybe there are other ways to fulfil your requirements that do not involve threads fiddling.

**Reply by akihiro-komorilab** [reply]
Thank you for your help.
I am looking to embed a subscriber implemented in zenoh-python into another application.
At this time, I want to control the start and end of the subscribe process from the application side without using any signals.
In other words, it would be great if you could tell me how to successfully stop a thread started by `declare_subscriber()`.

**Reply by Mallets** [reply]
A `declare_subscriber` does not start any thread. If you want to stop a subscription, you should call `subscription.undeclare()`. See [here](https://zenoh-python.readthedocs.io/en/latest/index.html?highlight=subscriber#subscriber).

**Reply by akihiro-komorilab** [reply]
Oh, really?
For example, why doesn't the process end when this [z_sub.py](https://raw.githubusercontent.com/eclipse-zenoh/zenoh-python/master/examples/z_sub.py) is changed and executed as follows?

```z_sub_mod.py
import sys
import time
from datetime import datetime
import argparse
import json
import zenoh
from zenoh import Reliability, Sample

# --- Command line argument parsing --- --- --- --- --- ---
parser = argparse.ArgumentParser(
    prog='z_sub',
    description='zenoh sub example')
parser.add_argument('--mode', '-m', dest='mode',
                    choices=['peer', 'client'],
                    type=str,
                    help='The zenoh session mode.')
parser.add_argument('--connect', '-e', dest='connect',
                    metavar='ENDPOINT',
                    action='append',
                    type=str,
                    help='Endpoints to connect to.')
parser.add_argument('--listen', '-l', dest='listen',
                    metavar='ENDPOINT',
                    action='append',
                    type=str,
                    help='Endpoints to listen on.')
parser.add_argument('--key', '-k', dest='key',
                    default='demo/example/**',
                    type=str,
                    help='The key expression to subscribe to.')
parser.add_argument('--config', '-c', dest='config',
                    metavar='FILE',
                    type=str,
                    help='A configuration file.')

args = parser.parse_args()
conf = zenoh.Config.from_file(
    args.config) if args.config is not None else zenoh.Config()
if args.mode is not None:
    conf.insert_json5(zenoh.config.MODE_KEY, json.dumps(args.mode))
if args.connect is not None:
    conf.insert_json5(zenoh.config.CONNECT_KEY, json.dumps(args.connect))
if args.listen is not None:
    conf.insert_json5(zenoh.config.LISTEN_KEY, json.dumps(args.listen))
key = args.key

# Zenoh code  --- --- --- --- --- --- --- --- --- --- ---



# initiate logging
zenoh.init_logger()

print("Opening session...")
session = zenoh.open(conf)

print("Declaring Subscriber on '{}'...".format(key))


def listener(sample: Sample):
    print(f">> [Subscriber] Received {sample.kind} ('{sample.key_expr}': '{sample.payload.decode('utf-8')}')")
    

# WARNING, you MUST store the return value in order for the subscription to work!!
# This is because if you don't, the reference counter will reach 0 and the subscription
# will be immediately undeclared.
sub = session.declare_subscriber(key, listener, reliability=Reliability.RELIABLE())

# print("Enter 'q' to quit...")
# c = '\0'
# while c != 'q':
#     c = sys.stdin.read(1)
#     if c == '':
#         time.sleep(1)
#
# # Cleanup: note that even if you forget it, cleanup will happen automatically when 
# # the reference counter reaches 0
# sub.undeclare()
# session.close()
```

I tested using Python 3.8.3 and zenoh-python 0.10.0rc0 on Windows 11.

**Reply by akihiro-komorilab** [reply]
Ah I see. If we don't call `sub.undeclare()`, the `sub` itself acquires the lock internally.

**Reply by p-avital** [reply]
Hi @akihiro-komorilab,

The issue here is due to Python's garbage collection methodology. Notice that the example code doesn't follow best practice in that it runs its code in global scope. The issue with that is that `sub` and `session` get declared in that global scope, which only gets cleaned up when all children threads have been joined.

However, this joining can only happen upon the session's destruction (note that subscribers hold a strong reference to the session, making their destruction necessary to fully closing a session), when the threads spawned by the underlying Rust library will be cleaned up.

This problem doesn't appear as soon as these `session` and `sub` variables are declared inside a scope (for example, a `main` function which gets called after checking `__name__ == "__main__"` as is best practice), since leaving that scope will trigger the automatic cleanup.

We have a few higher priorities at the moment, but we'll eventually switch the examples to use a `main` function, as well as document this a bit better.

Thanks for spotting this :)

**Reply by akihiro-komorilab** [reply]
Hi,

Thank you so much for the details. I appreciate your help. I understand.

I was able to get the information I needed to do what I wanted to do. I was also able to incorporate the zenoh-python module into another application, which is what I originally wanted to do.
I am currently compiling development information into a web article and will let you know when it is published. (in Japanese though ;)

Thanks again!

**Reply by p-avital** [reply]
良いな、ぜひリンクを教えてください!　（全然漢字を読めないんだけど...）

**Reply by akihiro-komorilab** [reply]
Sure! I will tell you tomorrow. (Your use of Kanji is perfect already!)

**Reply by akihiro-komorilab** [reply]
Hi @p-avital 
Here you are.
https://qiita.com/comocc/items/448deee5a0bf82b586ec
I had to mention you guys in the acknowledgements. :)
Thanks!
